classes/user/Collector.php

```diff
@@ -457,22 +457,14 @@ protected function buildUserGroupFilter(Builder $query): self
             ->when($this->roleIds !== null, fn ($subQuery) => $subQuery->whereIn('ug.role_id', $this->roleIds))
             ->when($this->contextIds !== null, fn ($subQuery) => $subQuery->whereIn('ug.context_id', $this->contextIds));
 
-        $joinSub = clone $subQuery;
-        $query
-            ->addWhereExistsQuery(
-                $subQuery->when(
-                    $this->excludeRoles != null,
-                    // This aggregates a column role_count, which holds an information whether user holds specified role
-                    fn (Builder $subQuery) => $subQuery->leftJoinSub(
-                        $joinSub
-                            ->select('uug.user_id')
-                            ->whereIn('ug.role_id', $this->excludeRoles),
-                        'agr',
-                        'uug.user_id',
-                        'agr.user_id'
-                    )->whereNull('agr.user_id')
-                )
-            );
+        $query->addWhereExistsQuery($subQuery);
+
+        if ($this->excludeRoles != null) {
+            $excludeRolesSubQuery = clone $subQuery;
+            $excludeRolesSubQuery->whereIn('ug.role_id', $this->excludeRoles);
+            $query->whereRaw("not exists({$excludeRolesSubQuery->toSql()})")
+                ->mergeBindings($excludeRolesSubQuery);
```

Comment:
I think you can avoid the `whereRaw()` if you move the `$subQuery` to a function.
```php
$getUserGroupsQuery = fn (Builder $q) => $q->from('user_user_groups as uug')->...;
$subQuery = $getUserGroupsQuery(DB::query());
...
$query->whereNotExists(fn (Builder $q) => $getUserGroupsQuery($q)->...)
```